### PR TITLE
alsoTo eager cancellation #24291

### PR DESCRIPTION
**Pull request by jeremystone**

### CODE REVIEW

**Comment by johanandren** [comment]
OK TO TEST

File: akka-docs/src/main/paradox/stream/stages-overview.md [hunk 1]
```diff
@@ -716,6 +716,8 @@ Attaches the given `Sink` to this `Flow`, meaning that elements that pass throug
 
 **completes** when upstream completes
 
+**Cancels** when downstream or `Sink` cancels
```
Comment (johanandren) [inline akka-docs/src/main/paradox/stream/stages-overview.md]: Nitpick: the other entries ones start with a lowercase letter

**Review by johanandren — APPROVED** [review]
LGTM

**Comment by raboof** [comment]
Refs #24291

**Comment by johanandren** [comment]
Seems there are a few tests relying on the old behavior of `alsoTo`.

**Comment by jeremystone** [comment]
@johanandren. Indeed.

The tests make use of alsoTo to force a behaviour that could be achieved by other means (at least by a raw Broadcast). But who else relies on this? 

Should I fix up the tests or do we (you guys) need to rethink and maybe have eagerCancel as a parameter of alsoTo?

**Comment by jeremystone** [comment]
OK. Will take a look at fixing up the tests when I get chance.

**Review by patriknw — APPROVED** [review]
LGTM

**Comment by jeremystone** [comment]
Please can someone check I haven't altered the semantics of the three tests I fixed that were affected by the changed behaviour of `alsoTo`. (Once `wireTap` (#23824) is merged in, these can probably be reverted to using that instead?)

**Review by johanandren — APPROVED** [review]
LGTM

**Comment by johanandren** [comment]
Sorry about us being slow here, I think this is ready for merge, but now needs a rebase first.

**Comment by johanandren** [comment]
Thanks!

File: akka-stream/src/main/scala/akka/stream/javadsl/SubSource.scala [hunk 2]
```diff
@@ -1137,7 +1137,7 @@ class SubSource[+Out, +Mat](delegate: scaladsl.SubFlow[Out, Mat, scaladsl.Source
    *
    * '''Completes when''' upstream completes
    *
-   * '''Cancels when''' downstream cancels
+   * '''Cancels when''' downstream or Sink cancels
    */
   def alsoTo(that: Graph[SinkShape[Out], _]): SubSource[Out, Mat] =
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/javadsl/SubSource.scala]: Could we have introduced an overload with the eager parameter?

Comment (patriknw) [reply akka-stream/src/main/scala/akka/stream/javadsl/SubSource.scala]: I think it was discussed and if you don't care about the side stream you should use the new wireTap (or graph dsl with Broadcast for power user full control)

Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/javadsl/SubSource.scala]: I'm not sure that's a good idea; that conflates the 2 decisions: 

- if the "other" should backpressure the upstream  (wireTap -- things get dropped for "other"; alsoTo -- things are backpressured)
- and what happens when "other" cancels (wireTap -- "other" completing does not cancel upstream; alsoTo -- completion of "other" does cancel the upstream)

But dropping on backpressure into "other" yet cancelling if it terminates could also be a valid use case?

**Pull request merged**